src/refiners/fluxion/layers/attentions.py

```diff
@@ -18,68 +18,137 @@ def scaled_dot_product_attention(
     value: Float[Tensor, "batch target_sequence_length dim"],
     is_causal: bool = False,
 ) -> Float[Tensor, "batch source_sequence_length dim"]:
-    return _scaled_dot_product_attention(query, key, value, is_causal=is_causal)  # type: ignore
+    """Scaled Dot Product Attention.
 
+    Optimization depends on which pytorch backend is used.
+    See [[arXiv:1706.03762] Attention Is All You Need (Equation 1)](https://arxiv.org/abs/1706.03762) for more details.
+    See also [torch.nn.functional.scaled_dot_product_attention](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html).
+    """
+    return _scaled_dot_product_attention(
+        query=query,
+        key=key,
+        value=value,
+        is_causal=is_causal,
+    )
 
-def sparse_dot_product_attention_non_optimized(
+
+def scaled_dot_product_attention_non_optimized(
     query: Float[Tensor, "batch source_sequence_length dim"],
     key: Float[Tensor, "batch target_sequence_length dim"],
     value: Float[Tensor, "batch target_sequence_length dim"],
     is_causal: bool = False,
 ) -> Float[Tensor, "batch source_sequence_length dim"]:
+    """Non-optimized Scaled Dot Product Attention.
+
+    See [[arXiv:1706.03762] Attention Is All You Need (Equation 1)](https://arxiv.org/abs/1706.03762) for more details.
+    """
     if is_causal:
         # TODO: implement causal attention
-        raise NotImplementedError("Causal attention for non_optimized attention is not yet implemented")
-    _, _, _, dim = query.shape
+        raise NotImplementedError(
+            "Causal attention for `scaled_dot_product_attention_non_optimized` is not yet implemented"
+        )
+
+    dim = query.shape[-1]
     attention = query @ key.permute(0, 1, 3, 2)
     attention = attention / math.sqrt(dim)
     attention = torch.softmax(input=attention, dim=-1)
     return attention @ value
 
 
 class ScaledDotProductAttention(Module):
+    """Scaled Dot Product Attention.
+
+    ??? note "See [[arXiv:1706.03762] Attention Is All You Need (Figure 2)](https://arxiv.org/abs/1706.03762) for more details"
```

Comment:
Is the `???` intended?

Reply:
Yes it's the syntax to show the note section collapsed. Let me see if I can find another syntax.